rust/rbac-registration/src/cardano/cip509/utils/extract_key.rs

```diff
@@ -1,19 +1,75 @@
 //! Functions for extracting public key from X509 and C509 certificates with additional
 //! verification.
 
-use anyhow::{anyhow, Context};
+use std::borrow::Cow;
+
 use c509_certificate::c509::C509;
-use ed25519_dalek::{VerifyingKey, PUBLIC_KEY_LENGTH};
+use catalyst_types::problem_report::ProblemReport;
+use ed25519_dalek::{SignatureError, VerifyingKey, PUBLIC_KEY_LENGTH};
 use oid_registry::{Oid, OID_SIG_ED25519};
 use thiserror::Error;
-use x509_cert::Certificate as X509Certificate;
+use x509_cert::{spki, Certificate as X509Certificate};
+
+/// Common error type.
+#[derive(Debug, Error)]
+pub enum Error {
+    /// Unsupported signature algorithms.
+    #[error("Unsupported signature algorithm (oid: {oid})")]
+    UnsupportedSignatureAlgo {
+        /// An OID of unsupported signature algorithm.
+        oid: Oid<'static>,
+    },
+    /// Public key has invalid length.
+    #[error(
+        "Invalid public key length (found {bits} bits, but expected {PUBLIC_KEY_LENGTH} bytes)"
+    )]
+    InvalidPublicKeyLength {
+        /// Number of the bits found.
+        ///
+        /// # Note
+        // Counting bits instead of bytes here, because of [`X509Certificate`] implementation of
+        // the public key storage.
+        bits: usize,
+    },
+    /// Public key doesn't pass [`ed25519_dalek`] constraint check.
+    #[error("Invalid public key ({source})")]
+    PublicKeyIsNotEd25519 {
+        /// Underlying [`ed25519_dalek`] error.
+        #[source]
```

Comment:
I rather use `from`, so it would allow to get rid of all the places with `map_err` and replace it with just `?`
```suggestion
        #[from]
```